File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nextGreater;
        stack<int> stk;  // monotonic decreasing stack

        for (int num : nums2) {
            while (!stk.empty() && stk.top() < num) {
                nextGreater[stk.top()] = num;
                stk.pop();
            }
            stk.push(num);
        }

        vector<int> result;
        for (int num : nums1) {
            result.push_back(nextGreater.count(num) ? nextGreater[num] : -1);
        }

        return result;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: list[int], nums2: list[int]) -> list[int]:
        next_greater = {}
        stack = []  # monotonic decreasing stack

        for num in nums2:
            # While current num is greater than the stack's top,
            # we've found the "next greater element" for that top
            while stack and stack[-1] < num:
                next_greater[stack.pop()] = num
            stack.append(num)

        # Elements remaining in the stack have no next greater element
        # (they're implicitly -1 via .get() default)

        return [next_greater.get(num, -1) for num in nums1]
        